keps/sig-apps/sidecarcontainers.md

```diff
@@ -156,9 +181,9 @@ Please view this [video](https://youtu.be/4hC8t6_8bTs) if you want to see what t
 
 ### Risks and Mitigations
 
-You could set all containers to be `sidecar: true`, this would cause strange behaviour in regards to shutting down the sidecars when all the non-sidecars have exited. To solve this the api could do a validation check that at least one container is not a sidecar.
+You could set all containers to be `containerLifecycle: Sidecar`, this would cause strange behaviour in regards to shutting down the sidecars when all the non-sidecars have exited. To solve this the api could do a validation check that at least one container is not a sidecar.
 
-Init containers would be able to have `sidecar: true` applied to them as it's an additional field to the container spec, this doesn't currently make sense as init containers are ran sequentially. We could get around this by having the api throw a validation error if you try to use this field on an init container or just ignore the field.
+Init containers would be able to have `containerLifecycle: Sidecar` applied to them as it's an additional field to the container spec, this doesn't currently make sense as init containers are ran sequentially. We could get around this by having the api throw a validation error if you try to use this field on an init container or just ignore the field.
```

Comment:
Not exactly your problem, but maybe we should just not be sharing the base type?

Reply:
e.g.

```
--- a/pkg/apis/core/types.go
+++ b/pkg/apis/core/types.go
@@ -1932,7 +1932,7 @@ type ResourceRequirements struct {
 }
 
 // Container represents a single container that is expected to be run on the host.
-type Container struct {
+type BaseContainer struct {
        // Required: This must be a DNS_LABEL.  Each container in a pod must
        // have a unique name.
        Name string
@@ -1993,6 +1993,9 @@ type Container struct {
        // If set, the fields of SecurityContext override the equivalent fields of PodSecurityContext.
        // +optional
        SecurityContext *SecurityContext
+}
+type Container struct {
+       BaseContainer // Embed
 
        // Variables for interactive containers, these have very specialized use-cases (e.g. debugging)
        // and shouldn't be used for general purpose containers.
@@ -2003,6 +2006,9 @@ type Container struct {
        // +optional
        TTY bool
 }
+type InitContainer struct {
+       BaseContainer // Embed
+}
```

This has wide-ranging code impact and probably openapi impact, but may be worth doing anyway...

Reply:
I don't think refactoring fields into a nested struct a level deeper is wire-compatible with our proto serializations

Reply:
This is the direction we're heading for ephemeral containers, but the primary motivator is to have separate docs so we can label some fields as "this doesn't work with ephemeral containers".

Separating fields would mean we wouldn't need to do this and an `EphemeralContainer` would only have fields allow for ephemeral containers, but complexity would increase with the differences between container types. Something like:

```
type BaseContainer struct {
    // All containers have a name
    Name string
    ...
}

type ServingContainer struct {
    // Only serving containers have ports, probes, etc.
    Ports []ContainerPorts
    ReadinessProbe *Probe
    Lifecycle *Lifecycle
    ...
}

type InteractiveContainer struct {
    TTY bool
    ...
}

type Container {
    BaseContainer
    ServingContainer
    InteractiveContainer
}

type InitContainer {
    BaseContainer
    ServingContainer
}

type EphemeralContainer {
    BaseContainer
    InteractiveContainer
}
```

It would be painful if we added a new container type or if we wanted to add a previously-disallowed field to EphemeralContainer or add a new container type that doesn't fit in these categories, especially considering @liggitt's point about proto serialization.